[inductor] Add API to make post_grad_custom passes cache-able

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137298

Summary: See https://github.com/pytorch/pytorch/issues/130772

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

masnesral committed:
[inductor] bypass FX graph cache if post_grade_custom passes are set

Summary: Disable caching until we have a proper solution. See https://github.com/pytorch/pytorch/issues/130772

[ghstack-poisoned]

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -1257,6 +1256,13 @@ def _check_can_cache(gm: torch.fx.GraphModule) -> None:
         to bypass in case caching is not possible.
         """
+        # TODO: https://github.com/pytorch/pytorch/issues/130772
+        if (
+            config.post_grad_custom_pre_pass is not None
+            or config.post_grad_custom_post_pass is not None
+        ):
+            raise BypassFxGraphCache("Post grad custom passes unsupported")
```
it'd be a couple more lines of code to check if `config.post_grad_custom_post_pass.get("fx_graph_cache_key")` exists so users could unblock themselves.  i think we should add it, or a variant

masnesral committed:
Update on "[RFC] [inductor] Add API to make post_grade_custom passes cache-able"


Summary: See https://github.com/pytorch/pytorch/issues/130772

[ghstack-poisoned]

masnesral commented:
@eellison lemme know if this looks like what you had in mind and I'll write the tests and fix the comments

eellison commented on torch/_inductor/post_grad_custom_pass.py:
```diff
@@ -0,0 +1,53 @@
+import torch.fx.graph
+
+
+class PostGradCustomPass(ABC):
```
Nit: Can we make this more generic of a name? There are also options for joint graph, and pre grad? CompilePass maybe ? idk

eellison commented on torch/_inductor/post_grad_custom_pass.py:
```diff
@@ -0,0 +1,53 @@
+
+    1) The __call__() method contains the implementation of the custom pass.
+
+    2) The uuid() method is necessary to support inductor caching compiled graphs when
```
nice doc !

eellison commented on torch/_inductor/post_grad_custom_pass.py:
```diff
@@ -0,0 +1,53 @@
+
+    2) The uuid() method is necessary to support inductor caching compiled graphs when
+    your custom passes are applied. This method can return any unique identifier, as
+    long as it uniquely identifies your custom pass implementation. The caching logic
```
nit: unique identifier so long as it can be pickled ?

eellison commented on torch/_inductor/post_grad_custom_pass.py:
```diff
@@ -0,0 +1,53 @@
+    @abstractmethod
+    def uuid(self) -> Union[bytes, str]:
+        """
+        Return an ID to uniquely identify your custom pass implementation. Return None
+        to bypass inductor code caching.
```
nit: not sure if bypass means cache miss, or bypass means assume the cache is valid.

oulgen replied:
bypass means cache miss, bypass the cache all together

eellison reviewed (approved):
Looks great ! just one comment on name

masnesral committed:
Update on "[inductor] Add API to make post_grad_custom passes cache-able"


Summary: See https://github.com/pytorch/pytorch/issues/130772

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Add API to make post_grad_custom passes cache-able"


Summary: See https://github.com/pytorch/pytorch/issues/130772

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Add API to make post_grad_custom passes cache-able"


Summary: See https://github.com/pytorch/pytorch/issues/130772

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Add API to make post_grad_custom passes cache-able"


Summary: See https://github.com/pytorch/pytorch/issues/130772

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
